### PR TITLE
M3 1127 Change: Rebuild from StackScript

### PR DESCRIPTION
**Pull request by johnwcallahan**
## Description

~TO BE REVIEWED AFTER https://github.com/linode/manager/pull/4593 IS MERGED.~
 **DONE**

Adds the ability to rebuild a Linode from a StackScript.

screen shot 2019-03-05 at 5 50 49 pm

## Type of Change
- Non breaking change ('update', 'change')

## Note to Reviewers

I experimented with a few Hooks patterns. I added 3 new utility hooks that can be used elsewhere: `useErrors`, `useForms`, and `useStackScript`. These are sort of WIP/POC-ish - feedback welcome. In particular, `useErrors` could be extended a lot – this PR represents a basic use-case.

**TO DO:**
- [x] Clean up `RebuildFromStackScript.tsx`
- [x] Add tests
- [x] Merge https://github.com/linode/manager/pull/4593

### CODE REVIEW

File: src/features/linodes/LinodesDetail/LinodeRebuild/RebuildFromStackScript.tsx [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        { field: 'stackscript_id', reason: 'You must select a StackScript' }
+      ]);
+      return;
+    }
+
+    rebuildLinode(linodeId, {
+      stackscript_id: ss.id,
+      stackscript_data: ss.udf_data,
+      root_pass: form.password,
+      image: form.imageID,
```
Comment (Jskobos) [inline src/features/linodes/LinodesDetail/LinodeRebuild/RebuildFromStackScript.tsx]: This should be cleared whenever the user selects a new stackscript. Currently the previous one is still selected, so the user can submit the form (so may get an incompatible image error or end up with a distro other than what they had intended)

File: src/hooks/useStackScript.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  user_defined_fields: [],
+  udf_data: []
+};
+
+const getCompatibleImages = (
+  allImages: Linode.Image[],
+  stackScriptImages: string[]
+) => {
+  return allImages.filter(image => {
+    for (const stackScriptImage of stackScriptImages) {
```
Comment (Jskobos) [inline src/hooks/useStackScript.ts]: I wonder if `stackScriptImages.includes(image.id)` would be faster than the double loop.

File: src/hooks/useStackScript.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      id,
+      username,
+      label,
+      images: getCompatibleImages(images, stackScriptImages),
+      user_defined_fields,
+      udf_data: getDefaultUDFData(user_defined_fields)
+    });
+  };
+
+  // Wrapper around setStackScript() – allows child components to handle change
```
Comment (Jskobos) [inline src/hooks/useStackScript.ts]: Comment doesn't match

**Comment by Jskobos** [comment]
Oh I forgot about this: 
screen shot 2019-03-06 at 4 02 23 pm

(when I block requests)

File: src/hooks/useStackScript.ts (hunk 3 above)
Comment (johnwcallahan) [reply src/hooks/useStackScript.ts]: 🤦‍♂️

File: src/hooks/useStackScript.ts (hunk 2 above)
Comment (johnwcallahan) [reply src/hooks/useStackScript.ts]: And cleaner too – thanks for the suggestion.

File: src/features/linodes/LinodesDetail/LinodeRebuild/RebuildFromStackScript.tsx (hunk 1 above)
Comment (johnwcallahan) [reply src/features/linodes/LinodesDetail/LinodeRebuild/RebuildFromStackScript.tsx]: Good call! Addressed.

**Pull request merged**